File: sdk/typescript/packages/cloud/src/cloud-pusher.ts
// CloudPusher — fire-and-forget Path B push for adapters in serverless / CI.
//
// The contract differs from the daemon's CloudClient (in jamjet-policy):
//   - Never throws. push() returns boolean. The agent's tool-call latency
//     budget MUST NOT be affected by Cloud reachability.
//   - Short timeout (default 500ms). A hung Cloud aborts quickly so the
//     adapter can keep serving its tool call.
//   - Circuit breaker: 5 consecutive failures opens the breaker for 60s.
//     During an open breaker, push() short-circuits to false without an
//     HTTP attempt — saves the 500ms budget on every subsequent call.
//   - 4xx and 5xx alike count as failures and return false. Direct-push has
//     no outbox to retry from, so anything other than a 2xx is dropped.
//     Operators wanting durable retry should run the sidecar daemon (Path A).
//
// Uses Node's built-in fetch (18+) so adapters don't pull in undici as a
// hard dependency. AbortController gives us a single primitive for both
// timeout and external cancellation.

export interface CloudPusherOptions {
  apiBase: string
  apiKey: string
  /** Default 500ms. */
  timeoutMs?: number
  /** Default 5. */
  circuitBreakerThreshold?: number
  /** Default 60_000. */
  circuitBreakerResetMs?: number
  userAgent?: string
}

// Loose shape — adapters can pass their own AuditEventV1-compatible objects.
// The Cloud API validates via its own audit-event-v1 schema, so we don't
// re-validate at this layer.
export interface CloudPusherEvent {
  ts: string
  run_id: string
  adapter: string
  host: string
  tool: string
  decision: string
  executed: boolean
  schema_version: 1
  args?: Record<string, unknown>
  args_redaction?: string
  trace_id?: string | null
  decision_id?: string | null
  rule?: string | null
  rule_kind?: string | null
  server?: string | null
  policy_version?: string
}

interface ResolvedOptions {
  apiBase: string
  apiKey: string
  timeoutMs: number
  circuitBreakerThreshold: number
  circuitBreakerResetMs: number
  userAgent: string
}

export class CloudPusher {
  consecutiveFailures = 0
  private circuitOpenedAt: number | undefined
  private readonly opts: ResolvedOptions

  constructor(opts: CloudPusherOptions) {
    this.opts = {
      apiBase: opts.apiBase,
      apiKey: opts.apiKey,
      timeoutMs: opts.timeoutMs ?? 500,
      circuitBreakerThreshold: opts.circuitBreakerThreshold ?? 5,
      circuitBreakerResetMs: opts.circuitBreakerResetMs ?? 60_000,
      userAgent: opts.userAgent ?? '@jamjet/cloud direct-push',
    }
  }

  isCircuitOpen(): boolean {
    if (this.circuitOpenedAt === undefined) return false
    if (Date.now() - this.circuitOpenedAt > this.opts.circuitBreakerResetMs) {
      this.circuitOpenedAt = undefined
      this.consecutiveFailures = 0
      return false
    }
    return true
  }

  async push(event: CloudPusherEvent): Promise<boolean> {
    if (this.isCircuitOpen()) return false

    const ac = new AbortController()
    const timer = setTimeout(() => ac.abort(), this.opts.timeoutMs)
    try {
      const resp = await fetch(`${this.opts.apiBase}/v1/policy-audit/events`, {
        method: 'POST',
        headers: {
          authorization: `Bearer ${this.opts.apiKey}`,
          'content-type': 'application/json',
          'user-agent': this.opts.userAgent,
        },
        body: JSON.stringify({ events: [event], path: 'direct' }),
        signal: ac.signal,
      })
      if (resp.ok) {
        this.consecutiveFailures = 0
        // Drain the body so the connection can be reused; otherwise Node
        // keeps the socket open until GC.
        try {
          await resp.text()
        } catch {
          // ignore drain failures
        }
        return true
      }
      this.recordFailure()
      return false
    } catch {
      this.recordFailure()
      return false
    } finally {
      clearTimeout(timer)
    }
  }

  private recordFailure(): void {
    this.consecutiveFailures++
    if (this.consecutiveFailures >= this.opts.circuitBreakerThreshold) {
      this.circuitOpenedAt = Date.now()
    }
  }
}

File: sdk/python/jamjet/cloud/cloud_pusher.py
"""CloudPusher — Path B fire-and-forget push for the Python SDK.

Mirror of the TS @jamjet/cloud CloudPusher contract:
  - Never raises. push() returns bool.
  - 500ms default timeout (httpx.Client(timeout=...)).
  - Circuit breaker: 5 consecutive failures → open for 60s.
  - 4xx and 5xx both count as failures. Direct-push has no outbox to retry
    from, so anything other than 2xx is dropped.

The agent's tool-call latency budget must not be affected by Cloud
reachability — that's the whole point of Path B's separation from the
local audit JSONL write.
"""

from __future__ import annotations

import os
import time
from typing import Any

import httpx


class CloudPusher:
    """Fire-and-forget Cloud push for Path B (serverless / CI)."""

    def __init__(
        self,
        api_base: str,
        api_key: str,
        *,
        timeout_seconds: float = 0.5,
        circuit_breaker_threshold: int = 5,
        circuit_breaker_reset_seconds: float = 60.0,
        user_agent: str = "jamjet python direct-push",
    ) -> None:
        self.api_base = api_base.rstrip("/")
        self.api_key = api_key
        self.timeout_seconds = timeout_seconds
        self.circuit_breaker_threshold = circuit_breaker_threshold
        self.circuit_breaker_reset_seconds = circuit_breaker_reset_seconds
        self.user_agent = user_agent
        self.consecutive_failures = 0
        self._circuit_opened_at: float | None = None
        self._client = httpx.Client(timeout=self.timeout_seconds)

    def is_circuit_open(self) -> bool:
        if self._circuit_opened_at is None:
            return False
        if time.time() - self._circuit_opened_at > self.circuit_breaker_reset_seconds:
            self._circuit_opened_at = None
            self.consecutive_failures = 0
            return False
        return True

    def push(self, event: dict[str, Any]) -> bool:
        """POST one audit event to /v1/policy-audit/events. Returns True on 2xx."""
        if self.is_circuit_open():
            return False
        try:
            response = self._client.post(
                f"{self.api_base}/v1/policy-audit/events",
                json={"events": [event], "path": "direct"},
                headers={
                    "authorization": f"Bearer {self.api_key}",
                    "user-agent": self.user_agent,
                },
            )
            if 200 <= response.status_code < 300:
                self.consecutive_failures = 0
                return True
            self._record_failure()
            return False
        except Exception:
            self._record_failure()
            return False

    def _record_failure(self) -> None:
        self.consecutive_failures += 1
        if self.consecutive_failures >= self.circuit_breaker_threshold:
            self._circuit_opened_at = time.time()

    def close(self) -> None:
        """Close the underlying httpx client. Idempotent."""
        try:
            self._client.close()
        except Exception:
            pass


SERVERLESS_ENV_VARS = (
    "VERCEL",
    "CF_PAGES",
    "AWS_LAMBDA_FUNCTION_NAME",
    "GITHUB_ACTIONS",
    "NETLIFY",
)


def detect_path_mode() -> str:
    """'local-only' or 'direct'. Mirror of the TS detectPathMode logic.

    Selection rules:
      1. No JAMJET_CLOUD_TOKEN → local-only.
      2. JAMJET_CLOUD_MODE=direct → direct.
      3. JAMJET_CLOUD_MODE=daemon → local-only.
      4. Any serverless indicator env var set → direct.
      5. Otherwise → local-only.
    """
    if not os.environ.get("JAMJET_CLOUD_TOKEN"):
        return "local-only"
    explicit = os.environ.get("JAMJET_CLOUD_MODE")
    if explicit == "direct":
        return "direct"
    if explicit == "daemon":
        return "local-only"
    for v in SERVERLESS_ENV_VARS:
        if os.environ.get(v):
            return "direct"
    return "local-only"
